zen3geo/datapipes/pystac_client.py

```diff
@@ -0,0 +1,136 @@
+        A DataPipe that contains STAC API query parameters in the form of a
+        Python dictionary to pass to :py:meth:`pystac_client.Client.search`.
+        For example:
+
+        - **bbox** -  A list, tuple, or iterator representing a bounding box of
+          2D or 3D coordinates. Results will be filtered to only those
+          intersecting the bounding box.
+        - **datetime** - Either a single datetime or datetime range used to
+          filter results. You may express a single datetime using a
+          :py:class:`datetime.datetime` instance, a
+          `RFC 3339-compliant <https://tools.ietf.org/html/rfc3339>`_
+          timestamp, or a simple date string.
+        - **collections** - List of one or more Collection IDs or
+          :py:class:`pystac.Collection` instances. Only Items in one of the
+          provided Collections will be searched.
+
+    catalog_url : str
+        The URL of a STAC Catalog. If not specified, this will use the
+        ``STAC_URL`` environment variable.
+
+    kwargs : Optional
+        Extra keyword arguments to pass to
+        :py:meth:`pystac_client.Client.open`. For example:
+
+        - **headers** - A dictionary of additional headers to use in all
+          requests made to any part of this Catalog/API.
+        - **parameters** - Optional dictionary of query string parameters to
+          include in all requests.
+        - **modifier** - A callable that modifies the children collection and
+          items returned by this Client. This can be useful for injecting
+          authentication parameters into child assets to access data from
+          non-public sources.
+
+    Yields
+    ------
+    item_search : pystac_client.ItemSearch
+        A :py:class:`pystac_client.ItemSearch` object instance that represents
+        a deferred query to a STAC search endpoint as described in the
+        `STAC API - Item Search spec <https://github.com/radiantearth/stac-api-spec/tree/main/item-search>`_.
+
+    Raises
+    ------
+    ModuleNotFoundError
+        If ``pystac_client`` is not installed. See
+        :doc:`install instructions for pystac-client <pystac_client:index>`,
+        (e.g. via ``pip install pystac-client``) before using this class.
+
+    Example
+    -------
+    >>> import pytest
+    >>> pystac_client = pytest.importorskip("pystac_client")
+    ...
+    >>> from torchdata.datapipes.iter import IterableWrapper
+    >>> from zen3geo.datapipes import PySTACAPISearch
+    ...
+    >>> # Peform STAC API query using DataPipe
+    >>> query = dict(
+    ...     bbox=[174.5, -41.37, 174.9, -41.19],
+    ...     datetime=["2012-02-20T00:00:00Z", "2022-12-22T00:00:00Z"],
+    ...     collections=["cop-dem-glo-30"],
+    ... )
+    >>> dp = IterableWrapper(iterable=[query])
+    >>> dp_pystac_client = dp.search_for_pystac_item(
+    ...     catalog_url="https://planetarycomputer.microsoft.com/api/stac/v1",
+    ...     # modifier=planetary_computer.sign_inplace,
```

Comment:
Commented out this `modifier` parameter (see also https://github.com/stac-utils/pystac-client/issues/259) because

1. It's a (cool) new feature done in https://github.com/stac-utils/pystac-client/pull/286, released in `pystac-client=0.4.7`, but we don't want to pin to a particular pystac-client version.
2. It requires the [`planetary-computer`](https://github.com/microsoft/planetary-computer-sdk-for-python) library to be installed.

So leaving this as an easter egg :egg: for now, but will include it in the next walkthrough tutorial.